plugins/targetlocker/mysql/mysql_target_locker.go

```diff
@@ -0,0 +1,448 @@
+
+	var mysqlError *mysql.MySQLError
+	if !xerrors.As(err, &mysqlError) {
+		return mysqlErrorTypeUnknown
+	}
+	switch mysqlError.Number {
+	case 1062:
+		return mysqlErrorTypeDuplicateKey
+	}
+	return mysqlErrorTypeOther
+}
+
+func (q mysqlLocksQuerier) actionLock(
+	mysql backendQuerier,
+	jobID types.JobID,
+	targetID string,
+) (result sql.Result, err error) {
+	// Remove expired an lock (if exists)
+
+	_, _ = q.actionClearExpiredLock(mysql, targetID)
+
+	// Add the lock
+
+	var errs []error
+
+	result, execErr := q.exec(mysql,
+		"INSERT INTO `"+q.tableName()+"` "+
+			"SET target_id=?, job_id=?, created_at = NOW(), expires_at = NOW() + INTERVAL ? SECOND",
+		targetID, jobID, int64(q.lockTTL.Seconds())) // #nosec G202
+
+	// Check errors:
+
+	if execErr != nil {
+		errs = append(errs, execErr)
+	}
+
+	if result != nil {
+		_, rowsAffectedErr := result.RowsAffected()
+		if rowsAffectedErr != nil {
+			errs = append(errs, rowsAffectedErr)
+		}
+	}
+
+	for _, errItem := range errs {
+		if q.errorType(errItem) == mysqlErrorTypeDuplicateKey {
+			err = targetlocker.ErrAlreadyLocked{
+				TargetID:        targetID,
+				UnderlyingError: errItem,
+			}
+		}
+	}
+
+	return
+}
+
+func (q mysqlLocksQuerier) actionUnlock(
+	mysql backendQuerier,
+	jobID types.JobID,
+	targetID string,
+) (sql.Result, error) {
+	return q.exec(mysql,
+		"DELETE FROM `"+q.tableName()+"` "+
+			"WHERE target_id=? AND job_id=?", targetID, jobID) // #nosec G202
+}
+
+func (q mysqlLocksQuerier) actionRefreshLock(
+	mysql backendQuerier,
+	jobID types.JobID,
+	targetID string,
+) (sql.Result, error) {
+	return q.exec(mysql,
+		"UPDATE `"+q.tableName()+"` "+
+			`SET expires_at = NOW() + INTERVAL ? SECOND WHERE target_id=? AND job_id=?`,
+		int64(q.lockTTL.Seconds()), targetID, jobID) // #nosec G202
+}
+
+func (q mysqlLocksQuerier) placeholder(arg interface{}) string {
+	switch arg := arg.(type) {
+	case []string:
+		if len(arg) == 0 {
+			return `()`
+		}
+		result := strings.Repeat(`?,`, len(arg))
+		return "(" + result[:len(result)-1] + ")"
+	default:
+		panic(fmt.Sprintf("unknown argument type: %T", arg))
+	}
+}
+
+func (q mysqlLocksQuerier) arguments(arg interface{}) (result []interface{}) {
+	switch arg := arg.(type) {
+	case []string:
+		for _, item := range arg {
+			result = append(result, item)
+		}
+	default:
+		panic(fmt.Sprintf("unknown argument type: %T", arg))
+	}
+	return
+}
+
+func (q mysqlLocksQuerier) getLocks(
+	mysql backendQuerier,
+	jobID types.JobID,
+	targetIDs []string,
+) (*sql.Rows, error) {
+	arguments := []interface{}{jobID}
+	arguments = append(arguments, q.arguments(targetIDs)...)
+	return q.query(mysql,
+		"SELECT `target_id` FROM `"+q.tableName()+"` "+
+			"WHERE job_id=? AND target_id IN "+q.placeholder(targetIDs),
+		arguments...) // #nosec G202
+}
+
+func (tl *TargetLocker) reportIfError(err error) {
+	if err == nil {
+		return
+	}
+	frame := runtimetools.Frame(1)
```

Comment:
If I do so, then I will see `reportIfError` in the logs, while I was to see the function which called `reportIfError`. And I don't want to write anything to logs if `err == nil`. But yeah, I think you right, it may be an overkill to implement `runtimetools` just because I was lazy to write `if err ==nil` if few places. I should remove it.

Reply:
On the other hand. This could be handy for other pieces of the code. I will agree with any solution. I think both ways are good: as it is now (more syntax sugar) or without `runtimetools` (less code).

Reply:
Added a comment.

Reply:
**Update:**
After discussions with Andrea, I will remove this method (and it's dependency `runtimetools`).

Reply:
Removed `runtimetools`.